wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs

```diff
@@ -0,0 +1,194 @@
+                    deriving (Functor, Applicative, Monad, MonadIO)
+
+instance HasLoggerName KeystoreM where
+    askLoggerName = return (LoggerName "Keystore")
+    modifyLoggerName _ action = action
+
+instance CanLog KeystoreM where
+    dispatchMessage _ln sev txt = logMessage sev txt
+
+{-------------------------------------------------------------------------------
+  Creating a keystore
+-------------------------------------------------------------------------------}
+
+-- | Creates a new keystore.
+newKeystore :: MonadIO m => FilePath -> m Keystore
+newKeystore fp = liftIO $ runIdentityT $ fromKeystore $ do
+    us <- takeUserSecret fp
+    Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+-- | Creates a legacy 'Keystore' by reading the 'UserSecret' from a 'NodeContext'.
+-- Hopefully this function will go in the near future.
+newLegacyKeystore :: MonadIO m => NodeContext -> m Keystore
+newLegacyKeystore ctx = do
+     us <- atomically $ readTVar $ ncUserSecret ctx
+     Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+-- | Creates a 'Keystore' out of a randomly generated temporary file (i.e.
+-- inside your $TMPDIR of choice). Suitable for testing.
+newInMemoryTransientKeystore :: MonadIO m => m Keystore
+newInMemoryTransientKeystore = liftIO $ runIdentityT $ fromKeystore $ do
+    tempDir         <- liftIO getTemporaryDirectory
+    (tempFile, hdl) <- liftIO $ openTempFile tempDir "keystore.key"
+    liftIO $ hClose hdl
+    us <- takeUserSecret tempFile
+    Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+
+-- | Release the resources associated with this 'Keystore'.
+-- This function is idempotent and can be called multiple times.
+releaseKeystore :: MonadIO m => Keystore -> m ()
+releaseKeystore (Keystore ks) = liftIO $ modifyMVar ks $ \(InternalStorage us st) ->
+    case st of
+         StorageReleased      -> return ((InternalStorage us StorageReleased), ())
+         StorageInitialised   -> do
+             writeUserSecretRelease us
+             return ((InternalStorage us StorageReleased), ())
+
+
+{-------------------------------------------------------------------------------
+  Destroying a keystore
+-------------------------------------------------------------------------------}
+
+-- | Destroys a 'Keystore'.
+-- Completely obliterate the keystore from disk, with all its secrets.
+-- This operation cannot be reverted.
+-- This is a very destructive option that most of the time you
+-- probably don't want, as it will effectively leave you with an empty
+-- 'MVar' and your next call to an operation on the keystore would block
+-- indefinitely.
+-- This has still its use in some teardowns or tests.
+-- Note that this operation will always succeed. Use with care.
+destroyKeystore :: MonadIO m => Keystore -> m ()
+destroyKeystore (Keystore ks) = liftIO $ do
+    (InternalStorage us _st) <- takeMVar ks
+    writeUserSecretRelease us
+    removeFile (getUSPath us)
```

Comment:
Leaving this in an empty `MVar` state is not a good idea, if this ever happens the code will break horribly with an opaque "blocked indefinitely" exception. Moreover, there is code duplication here. Let's call `releaseKeystore` first, avoiding code duplication, and then remove the file.